sw/device/tests/alert_handler_lpg_clkoff_test.c

```diff
@@ -0,0 +1,743 @@
+    for_each_peripheral
+      1- wait for random time
+      2- Disable the peripheral's clock
+      3- Wait long enough
+      4- Confirm that ping_timeout has not been triggerred
+  */
+  if (test_step_cnt == 1 * ARRAYSIZE(kPeripherals)) {
+    alert_handler_config_all_peripherals(/* ping_timeout*/ 256);
+  }
+
+  while (test_step_cnt < 2 * ARRAYSIZE(kPeripherals) - 1) {
+    // Each test phase consists of ARRAYSIZE(kPeripherals) steps
+    // At every step, a specific peripheral is tested
+    test_step_cnt = flash_ctrl_testutils_counter_get(kCounterTestSteps);
+    test_phase = test_step_cnt / ARRAYSIZE(kPeripherals);
+    peri_idx = test_step_cnt - (test_phase)*ARRAYSIZE(kPeripherals);
+    // LOG_INFO("Test phase = %d, Periphereal idx = %d, test_step_cnt = %d",
+    // test_phase, peri_idx, test_step_cnt);
+
+    busy_spin_micros(rnd_wait_times[peri_idx]);
+
+    // Disable the clock of the peripheral
+    if (kPeripherals[peri_idx].is_hintable) {
+      CHECK_DIF_OK(dif_clkmgr_hintable_clock_set_hint(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleDisabled));
+    } else {
+      CHECK_DIF_OK(dif_clkmgr_gateable_clock_set_enabled(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleDisabled));
+    }
+
+    // wait enough
+    if (kDeviceType == kDeviceFpgaCw310) {
+      // NUM_ALERTS*2*margin_of_safety*(2**DW)*(1/kClockFreqPeripheralHz)
+      // (2**6)*2*4*(2**16)*(400ns) = 12.8s
+      busy_spin_micros(1000 * 12800);
+    } else if (kDeviceType == kDeviceSimDV) {
+      // NUM_ALERTS*2*margin_of_safety*(2**DW)*(1/kClockFreqPeripheralHz)
+      // (2**6)*2*4*(2**3)*(40ns) = 160us
+      busy_spin_micros(160);
+    }
+
+    // Check the alert status
+    dif_alert_handler_local_alert_t check_local_alert =
+        kDifAlertHandlerLocalAlertAlertPingFail;
+    CHECK_DIF_OK(dif_alert_handler_local_alert_is_cause(
+        &alert_handler, check_local_alert, &is_cause));
+    CHECK(!is_cause, "Expected response: No alert_ping_fail! but we got %d",
+          is_cause);
+
+    if (kPeripherals[peri_idx].is_hintable) {
+      CHECK_DIF_OK(dif_clkmgr_hintable_clock_set_hint(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleEnabled));
+    } else {
+      CHECK_DIF_OK(dif_clkmgr_gateable_clock_set_enabled(
+          &clkmgr, kPeripherals[peri_idx].clk_index, kDifToggleEnabled));
+    }
+
+    // Increment the test counter
+    flash_ctrl_testutils_counter_increment(&flash_ctrl, kCounterTestSteps);
+
+    // Reset the device after last peripheral
+    // to unlock the alert handler for the next phase
+    if (test_step_cnt == (test_phase + 1) * ARRAYSIZE(kPeripherals) - 1) {
+      CHECK_DIF_OK(dif_rstmgr_software_device_reset(&rstmgr));
+    }
+  }
+
+  /* TEST PHASE #2: TEST THE FATAL ALERTS
+
+    The timeout value is set to 256
+    None of the alerts should trigger ping_timeout_alert
+    TEST:
+    enable_alerts (all_peripherals, timeout=256)
+    for_each_peripheral
```

Comment:
This sequence probably needs some modifications.

The problem is that writing to an alert test register will only trigger one alert event, regardless of whether an alert is actually fatal or not. The reason for that is that SW cannot accidentally trigger a fatal alert that keeps on sending alert events until reset.

This means that we will potentially have to use DV backdoor forces in order to trigger fatal alerts.

I think the test currently passes only because we do not clear the cause bit on the alert handler side before re-enabling the peripheral clock.

@cindychip WDYT?

Reply:
Agree with Michael on alert_test register. If we want to check fatal alert, writing to `alert_test` register might not be sufficient.
I would recommend look at this test: 
https://github.com/lowRISC/opentitan/blob/master/hw/top_earlgrey/dv/env/seq_lib/chip_sw_all_escalation_resets_vseq.sv
It forces some internal prim_onehot error to trigger a fatal alert.

I am not aware of any fatal alert that can be triggered by SW only..
But with this check, this test will become a DV only check..

Reply:
Thanks for the feedback and the pointer to the DV sequence. I checked the file but I have a few questions to understand the DV side better. I will set up a meeting with Cindy.

Reply:
Thanks everyone for the feedback! As we discussed in the meeting - let us remove this part so the current C test can run in FPGA, DV, and Verilator.
We can leave a TODO here and create a separate C + DV test to force a fatal alert and verify the last part.
Then let us link both tests in the chip_testplan.

Reply:
Thanks everyone! I applied the feedback, and cleaned the code little bit more. I removed the fatal-alert part. To reduce the test time in the simulation, I also made the negative testing phase FPGA-only, which takes ~100s on the FPGA.